### PR TITLE
Installation problem with MPI on Windows

### PR DESCRIPTION
**Pull request by DaneilSun**
After these changes, step-31 can run successfully on Windows

### CODE REVIEW

File: include/deal.II/fe/fe_tools_interpolate.templates.h [hunk 1]
```diff
@@ -738,8 +738,13 @@ namespace FETools
       interpolation_difference(dof1, u1, dof2.get_fe(), u1_difference);
     else
       {
+#ifndef _MSC_VER
         internal::interpolation_difference(
           dof1, constraints1, u1, dof2, constraints2, u1_difference);
+#else
+        internal::interpolation_difference<dim, InVector, OutVector, spacedim>(
+          dof1, constraints1, u1, dof2, constraints2, u1_difference);
+#endif
```
Comment (marcfehling) [inline include/deal.II/fe/fe_tools_interpolate.templates.h:747]:
```suggestion
        internal::interpolation_difference<dim, InVector, OutVector, spacedim>(
          dof1, constraints1, u1, dof2, constraints2, u1_difference);
```

File: include/deal.II/grid/grid_tools.h [hunk 2]
```diff
@@ -188,13 +188,23 @@ namespace GridTools
...
          .template get_default_linear_mapping<dim, spacedim>()));
+#else
+  template <int dim, int spacedim>
+  double
+  minimal_cell_diameter(
+    const Triangulation<dim, spacedim> &triangulation,
+    const Mapping<dim, spacedim> &mapping =
+      (ReferenceCells::get_hypercube<dim>()
+         .ReferenceCell::get_default_linear_mapping<dim, spacedim>()));
+#endif
```
Comment (marcfehling) [inline include/deal.II/grid/grid_tools.h:207]:
```suggestion
  template <int dim, int spacedim>
  double
  minimal_cell_diameter(
    const Triangulation<dim, spacedim> &triangulation,
    const Mapping<dim, spacedim> &mapping =
      (ReferenceCells::get_hypercube<dim>()
         .ReferenceCell::get_default_linear_mapping<dim, spacedim>()));
```

File: include/deal.II/grid/grid_tools.h [hunk 3]
```diff
@@ -4358,11 +4371,21 @@ namespace GridTools
...
       });
+#    else
+      process_cells([&](const auto &cell, const auto key) -> void {
+        if (cell_filter(cell))
+          {
+            neighbor_cell_list[key].emplace_back(
+              cell->id().template to_binary<spacedim_template>());
+          }
+      });
+#    endif
```
Comment (marcfehling) [inline include/deal.II/grid/grid_tools.h:4388]:
```suggestion
      process_cells([&](const auto &cell, const auto key) -> void {
        if (cell_filter(cell))
          {
            constexpr int spacedim = MeshType::space_dimension;
            neighbor_cell_list[key].emplace_back(
              cell->id().template to_binary<spacedim>());
          }
      });
```

File: include/deal.II/grid/grid_tools.h [hunk 4]
```diff
@@ -4323,9 +4333,12 @@ namespace GridTools
...
 #  else
-      constexpr int dim      = MeshType::dimension;
-      constexpr int spacedim = MeshType::space_dimension;
-      auto          tria =
+      constexpr int    dim               = MeshType::dimension;
+      constexpr int    spacedim          = MeshType::space_dimension;
+#    ifdef _MSC_VER
+      static const int spacedim_template = MeshType::space_dimension;
+#    endif
+      auto             tria =
```
Comment (marcfehling) [inline include/deal.II/grid/grid_tools.h:4341]:
```suggestion
      constexpr int dim      = MeshType::dimension;
      constexpr int spacedim = MeshType::space_dimension;
      auto          tria =
```

File: include/deal.II/numerics/vector_tools_constraints.h [hunk 5]
```diff
@@ -305,6 +306,18 @@ namespace VectorTools
...
+  void
+  compute_no_normal_flux_constraints(
+    const DoFHandler<dim, spacedim> &   dof_handler,
+    const unsigned int                  first_vector_component,
+    const std::set<types::boundary_id> &boundary_ids,
+    AffineConstraints<double> &         constraints,
+    const Mapping<dim, spacedim> &      mapping =
+      (ReferenceCells::get_hypercube<dim>()
+         .ReferenceCell::get_default_linear_mapping<dim, spacedim>()));
+#endif
```
Comment (marcfehling) [inline include/deal.II/numerics/vector_tools_constraints.h:320]: Just keep the MSVC compatible variant, and remove the original one.

**Review by marcfehling — CHANGES_REQUESTED** [review]
Thank you for your contribution, and welcome to the group of deal.II contributors :)

First, I see you already opened a pull request #13592 for this. You do not need to close and open PRs again, you can simply work on the corresponding branch and things will get updated here.

Second, could you please provide your firstname and lastname to your git configuration? This way, we can properly credit you in our list of deal.II contributors. Please follow this [guide](https://github.com/dealii/dealii/wiki/Commit-authorship) (if you haven't already, as this was an issue in your old pull request).

---

Now about the content of your pull request:

With which version of MSVC does your problem occur? Our testsuite uses `MSVC 19.31`. I'm surprised it didn't catch this.

I would suggest to avoid conditionals based on the compiler when it's avoidable, especially on the interface in the header files. Your suggestions for the MSVC compiler should all be compatible to other compilers as well, so I would suggest to simply replace the original code. I've added some suggestions so that you can see what I mean. Could you give it a try?

**Comment by DaneilSun** [comment]
> Thank you for your contribution, and welcome to the group of deal.II contributors :)
> 
> First, I see you already opened a pull request #13592 for this. You do not need to close and open PRs again, you can simply work on the corresponding branch and things will get updated here.
> 
> Second, could you please provide your firstname and lastname to your git configuration? This way, we can properly credit you in our list of deal.II contributors. Please follow this [guide](https://github.com/dealii/dealii/wiki/Commit-authorship) (if you haven't already, as this was an issue in your old pull request).
> 
> Now about the content of your pull request:
> 
> With which version of MSVC does your problem occur? Our testsuite uses `MSVC 19.31`. I'm surprised it didn't catch this.
> 
> I would suggest to avoid conditionals based on the compiler when it's avoidable, especially on the interface in the header files. Your suggestions for the MSVC compiler should all be compatible to other compilers as well, so I would suggest to simply replace the original code. I've added some suggestions so that you can see what I mean. Could you give it a try?

 Yes, I will try it. Thanks for your kind suggestions. 

 I use 'MSVC 19.29', and the building fails with the MPI option = ON.

**Comment by marcfehling** [comment]
> I use 'MSVC 19.29', and the building fails with the MPI option = ON.
>
> The "interpolation_difference" function problem is caused by ambiguous call to overloaded function introduced by Trilinos.

Our CI does not check MSVC with neither MPI nor Trilinos. This might be why we didn't see the problem earlier.

> After changing the lambda functions' problem, building process was ok with the 'template' keyword. But when you run step-31, MSVC compiler complained about the invalid template syntax. So these places need to be fixed.

Our Windows CI also does not compile the example problems.

So these all are shortcomings in our CI. We need to update it....

**Comment by marcfehling** [comment]
`gcc` is fine with the new syntax, however `clang` complains about the missing `template` keyword.

I will push one commit to your branch that recovers your changes, but in a slim version. Let's see how the CI will react.

**Comment by marcfehling** [comment]
This is related to #11604.

**Comment by DaneilSun** [comment]
> This is related to #11604.
> 
> [6491133](https://github.com/dealii/dealii/commit/6491133d1ce5183ee58b76ecc8bc383c907a3866) shows what other places could be affected.

The fact is that building process goes well with "template" keyword. But when your codes run, MSVC complier is unhappy about the "template" keyword which it encounters.

**Comment by marcfehling** [comment]
Looks like the testsuite runs fine!

We still need to consider a few things:
- [ ] Can you verify that `deal.II` compiles on your machine with these changes?
It looks like you and @nicola-giuliani (in #12263) were the only ones who encountered it and can reproduce it. 
- [ ] Could you add a changelog entry in the `doc/news/changes/minor` folder?
You can take over the layout from one of the other files there and also find some inspiration for your changelog message.
- [ ] Would you be willing to update the other occurrences of the `template` keyword as well?
They are all documented in https://github.com/dealii/dealii/commit/6491133d1ce5183ee58b76ecc8bc383c907a3866 .
- [ ] Squash all commits of your feature branch into one commit

**Comment by peterrum** [comment]
@DaneilSun Could you address @marcfehling's comments?

**Review by marcfehling — APPROVED** [review]
Looks good to me!

I do not have a Windows system at my disposal to check this patch. If @DaneilSun says it works for him, then this is okay with me.

**Review by peterrum — APPROVED** [review]

**Pull request merged**